templates/fine-tune-llama2/train.py

```diff
@@ -0,0 +1,776 @@
+        type=str,
+        default="./deepspeed_configs/zero_3_llama_2_7b.json",
+        help="Deepspeed config json to use.",
+    )
+
+    parser.add_argument(
+        "--lora",
+        action="store_true",
+        default=False,
+        help="If passed, will enable parameter efficient fine-tuning with LoRA ("
+        "https://arxiv.org/pdf/2106.09685.pdf).",
+    )
+
+    parser.add_argument(
+        "--size",
+        default="7b",
+        choices=["7b", "13b", "70b"],
+        help="Model size. Choose between 7b, 13b, and 70b",
+    )
+
+    args = parser.parse_args()
+
+    return args
+
+
+def main():
+    print("Fine-tuning model...")
+    args = parse_args()
+
+    # Create the config with args for training.
+    config = vars(args)
+
+    SIZE = args.size # Variable to store the model size
+    LR = 5e-6 # Default learning rate for full-parameter fine-tuning
+
+    # Adjust learning rate if LoRA is enabled
+    if args.lora:
+        LR = 1e-4
+
+    # Adjust batch size per device (BS) and number of devices (ND) according to model size
+    # Number of devices (ND) is determined by a combination of factors, context length, accelerator type, whether LoRA is used, etc.
+    if SIZE == "7b" or SIZE == "13b":  
+        # ND is et on the basis of using Nvidia A10 and conducting full parameter fine-tuning with default context length.
+        # If Nvidia A100 is used. ND can be set to 8 instead.
+        BS, ND = 16, 16
+    elif SIZE == "70b":
+        # ND is set on the basis of using Nvidia A10 and conducting full parameter fine-tuning with default context length.
+        # If Nvidia A100 is used. ND can be set to 16 instead.
+        BS, ND = 8, 32
+    else:
+        print(f"Invalid size: {SIZE}")
+        sys.exit(1)
+
+    # Pick model ID and deepspeed config based on model size
+    MODEL_ID = f"meta-llama/Llama-2-{SIZE}-hf"
+    CONFIG_DIR = f"./deepspeed_configs/zero_3_llama_2_{SIZE}.json"
+
+    # Update the config with the adjusted parameters
+    config.update(
+        **{
+            "as_test": args.as_test,
+            "batch_size": BS,
+            "batch_size_per_device": args.batch_size_per_device,
+            "block_size": args.ctx_len,
+            "ds_config": CONFIG_DIR,
+            "eval_batch_size": args.eval_batch_size_per_device,
+            "grad_accum": args.grad_accum,
+            "lora": args.lora,
+            "lr": LR,
+            "mixed_precision": args.mixed_precision,
+            "model_name": MODEL_ID,
+            "no_grad_ckpt": args.no_grad_ckpt,
+            "num_checkpoints_to_keep": args.num_checkpoints_to_keep,
+            "num_devices": ND,
+            "num_epochs": args.num_epochs,
+            "output_dir": args.output_dir,
+            "seed": 42,
+            "special_token_path": args.special_token_path,
+            "stop_perplexity": args.stop_perplexity,
+            "train_path": args.train_path,
+            "test_path": args.test_path,
+        }
+    )
```

Comment:
Let's follow the common Python convention that only constants are fully in capitalized letters.
The capital letters are an artifact of this coming from a bash script.

Let's run all model sizes once with LoRA enabled please to verify that this works.

```suggestion
    size = args.size # Variable to store the model size
    lr = 5e-6 # Default learning rate for full-parameter fine-tuning

    # Adjust learning rate if LoRA is enabled
    if args.lora:
        lr = 1e-4

    # Adjust batch size per device (BS) and number of devices (ND) according to model size
    # Number of devices (ND) is determined by a combination of factors, context length, accelerator type, whether LoRA is used, etc.
    if size == "7b" or size == "13b":  
        # nd is set on the basis of using Nvidia A10 and conducting full parameter fine-tuning with default context length.
        # If Nvidia A100 is used. ND can be set to 8 instead.
        bs, nd = 16, 16
    elif size == "70b":
        # nd is set on the basis of using Nvidia A10 and conducting full parameter fine-tuning with default context length.
        # If Nvidia A100 is used. ND can be set to 16 instead.
        bs, nd = 8, 32
    else:
        print(f"Invalid size: {size}")
        sys.exit(1)

    # Pick model ID and deepspeed config based on model size
    model_id = f"meta-llama/Llama-2-{size}-hf"
    config_dir = f"./deepspeed_configs/zero_3_llama_2_{size}.json"

    # Update the config with the adjusted parameters
    config.update(
        **{
            "as_test": args.as_test,
            "batch_size": bs,
            "batch_size_per_device": args.batch_size_per_device,
            "block_size": args.ctx_len,
            "ds_config": config_dir,
            "eval_batch_size": args.eval_batch_size_per_device,
            "grad_accum": args.grad_accum,
            "lora": args.lora,
            "lr": lr,
            "mixed_precision": args.mixed_precision,
            "model_name": model_id,
            "no_grad_ckpt": args.no_grad_ckpt,
            "num_checkpoints_to_keep": args.num_checkpoints_to_keep,
            "num_devices": nd,
            "num_epochs": args.num_epochs,
            "output_dir": args.output_dir,
            "seed": 42,
            "special_token_path": args.special_token_path,
            "stop_perplexity": args.stop_perplexity,
            "train_path": args.train_path,
            "test_path": args.test_path,
        }
    )
```

Reply:
Sorry I didn't catch this earlier.